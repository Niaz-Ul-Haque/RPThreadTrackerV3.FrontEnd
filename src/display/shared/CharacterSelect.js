import React from 'react';
import { FormGroup, Label, Input } from 'reactstrap';
import PropTypes from 'prop-types';

const propTypes = {
	characters: PropTypes.arrayOf(PropTypes.shape({})).isRequired,
	selectedCharacterId: PropTypes.number,
	onSelectCharacter: PropTypes.func.isRequired,
	defaultText: PropTypes.string,
	includeNullValue: PropTypes.bool
};
const defaultProps = {
	selectedCharacterId: -1,
	defaultText: 'Select Character',
	includeNullValue: true
};

const CharacterSelect = (props) => {
	const {
		characters, selectedCharacterId, onSelectCharacter, defaultText, includeNullValue
	} = props;
	const options = [];
	if (characters) {
		for (let i = 0; i < characters.length; i++) {
			const element = (
				<option
					value={characters[i].characterId}
					key={characters[i].characterId}
				>
					{characters[i].characterName ? characters[i].characterName : characters[i].urlIdentifier}
				</option>
			);
			options.push(element);
		}
	}
	return (
		<FormGroup>
			<Label htmlFor="characterId">Character</Label>
			<Input
				type="select"
				name="characterId"
				id="character-id"
				defaultValue={selectedCharacterId}
				onChange={e => onSelectCharacter(parseInt(e.target.value, 10))}
			>
				{includeNullValue && <option value={null}>{defaultText}</option>}
				{options}
			</Input>
		</FormGroup>
	);
};

CharacterSelect.propTypes = propTypes;
CharacterSelect.defaultProps = defaultProps;
export default CharacterSelect;
