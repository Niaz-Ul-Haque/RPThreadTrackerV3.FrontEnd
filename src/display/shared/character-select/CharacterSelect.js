// #region imports
import React from 'react';
import { FormGroup, Label, Input } from 'reactstrap';
import PropTypes from 'prop-types';
import CharacterSelectItem from './CharacterSelectItem';
// #endregion imports

const propTypes = {
	characters: PropTypes.arrayOf(PropTypes.shape({})).isRequired,
	selectedCharacterId: PropTypes.number,
	onSelectCharacter: PropTypes.func.isRequired,
	defaultText: PropTypes.string,
	includeNullValue: PropTypes.bool
};
const defaultProps = {
	selectedCharacterId: -1,
	defaultText: 'Select Character',
	includeNullValue: true
};

const CharacterSelect = (props) => {
	const {
		characters, selectedCharacterId, onSelectCharacter, defaultText, includeNullValue
	} = props;
	const options = characters.map(c => <CharacterSelectItem character={c} />);
	return (
		<FormGroup>
			<Label htmlFor="characterId">Character</Label>
			<Input
				type="select"
				name="characterId"
				id="character-id"
				defaultValue={selectedCharacterId}
				onChange={e => onSelectCharacter(parseInt(e.target.value, 10))}
			>
				{includeNullValue && <option value={null}>{defaultText}</option>}
				{options}
			</Input>
		</FormGroup>
	);
};

CharacterSelect.propTypes = propTypes;
CharacterSelect.defaultProps = defaultProps;
export default CharacterSelect;
