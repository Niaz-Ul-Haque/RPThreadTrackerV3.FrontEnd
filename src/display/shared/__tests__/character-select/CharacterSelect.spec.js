// #region imports
import React from 'react';
import { shallow } from 'enzyme';
import { getSpecWrapper } from '../../../../../config/tests/helpers.unit';
import CharacterSelect from '../../character-select/CharacterSelect';
// #endregion imports

jest.mock('../../character-select/CharacterSelectItem', () => 'CharacterSelectItem');

const createTestProps = propOverrides => ({
	characters: [],
	onSelectCharacter: jest.fn(),
	...propOverrides
});
describe('rendering', () => {
	describe('snapshots', () => {
		it('should render valid snapshot', () => {
			const props = createTestProps();
			const jsx = (<CharacterSelect {...props} />);
			const element = shallow(jsx);
			expect(element).toMatchSnapshot();
		});
		it('should render valid snapshot with items', () => {
			const props = createTestProps({
				characters: [
					{ characterId: 1 },
					{ characterId: 2 },
					{ characterId: 3 }
				]
			});
			const jsx = (<CharacterSelect {...props} />);
			const element = shallow(jsx);
			expect(element).toMatchSnapshot();
		});
		it('should render valid snapshot with items and no null value', () => {
			const props = createTestProps({
				characters: [
					{ characterId: 1 },
					{ characterId: 2 },
					{ characterId: 3 }
				],
				includeNullValue: false
			});
			const jsx = (<CharacterSelect {...props} />);
			const element = shallow(jsx);
			expect(element).toMatchSnapshot();
		});
		it('should render valid snapshot with non-default props', () => {
			const props = createTestProps({
				selectedCharacterId: 5,
				defaultText: 'Select a Character'
			});
			const jsx = (<CharacterSelect {...props} />);
			const element = shallow(jsx);
			expect(element).toMatchSnapshot();
		});
	});
	describe('initial load', () => {
		it('should display blank when no characters', () => {
			const props = createTestProps({
				characters: []
			});
			const jsx = (<CharacterSelect {...props} />);
			const element = shallow(jsx);
			const items = element.find('CharacterSelectItem');
			expect(items).toHaveLength(0);
		});
		it('should display characters', () => {
			const props = createTestProps({
				characters: [
					{ characterId: 1 },
					{ characterId: 2 },
					{ characterId: 3 }
				]
			});
			const jsx = (<CharacterSelect {...props} />);
			const element = shallow(jsx);
			const items = element.find('CharacterSelectItem');
			const nullValues = getSpecWrapper(element, 'character-select-null-value');
			expect(items).toHaveLength(3);
			expect(nullValues).toHaveLength(1);
		});
		it('should display characters without null value', () => {
			const props = createTestProps({
				characters: [
					{ characterId: 1 },
					{ characterId: 2 },
					{ characterId: 3 }
				],
				includeNullValue: false
			});
			const jsx = (<CharacterSelect {...props} />);
			const element = shallow(jsx);
			const items = element.find('CharacterSelectItem');
			const nullValues = getSpecWrapper(element, 'character-select-null-value');
			expect(items).toHaveLength(3);
			expect(nullValues).toHaveLength(0);
		});
	});
});

describe('behavior', () => {
	describe('onSelectCharacter', () => {
		it('should be called when field value changes', () => {
			const onSelectCharacter = jest.fn();
			const mockEvent = { target: { value: 10 } };
			const props = createTestProps({ onSelectCharacter });
			const jsx = (<CharacterSelect {...props} />);
			const element = shallow(jsx);
			const field = getSpecWrapper(element, 'character-select-field');
			field.simulate('change', mockEvent);
			expect(onSelectCharacter).toHaveBeenCalledTimes(1);
			expect(onSelectCharacter).toHaveBeenLastCalledWith(10);
		});
	});
});
