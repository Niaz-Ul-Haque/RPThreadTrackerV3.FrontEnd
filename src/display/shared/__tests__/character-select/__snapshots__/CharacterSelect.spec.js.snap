// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering should render valid snapshot 1`] = `
<FormGroup
  tag="div"
>
  <Label
    htmlFor="characterId"
    tag="label"
  >
    Character
  </Label>
  <Input
    data-spec="character-select-field"
    defaultValue={-1}
    id="character-id"
    name="characterId"
    onChange={[Function]}
    tag="p"
    type="select"
  >
    <option
      data-spec="character-select-null-value"
      value={null}
    >
      Select Character
    </option>
  </Input>
</FormGroup>
`;

exports[`rendering should render valid snapshot with items 1`] = `
<FormGroup
  tag="div"
>
  <Label
    htmlFor="characterId"
    tag="label"
  >
    Character
  </Label>
  <Input
    data-spec="character-select-field"
    defaultValue={-1}
    id="character-id"
    name="characterId"
    onChange={[Function]}
    tag="p"
    type="select"
  >
    <option
      data-spec="character-select-null-value"
      value={null}
    >
      Select Character
    </option>
    <CharacterSelectItem
      character={
        Object {
          "characterId": 1,
        }
      }
      key="1"
    />
    <CharacterSelectItem
      character={
        Object {
          "characterId": 2,
        }
      }
      key="2"
    />
    <CharacterSelectItem
      character={
        Object {
          "characterId": 3,
        }
      }
      key="3"
    />
  </Input>
</FormGroup>
`;

exports[`rendering should render valid snapshot with items and no null value 1`] = `
<FormGroup
  tag="div"
>
  <Label
    htmlFor="characterId"
    tag="label"
  >
    Character
  </Label>
  <Input
    data-spec="character-select-field"
    defaultValue={-1}
    id="character-id"
    name="characterId"
    onChange={[Function]}
    tag="p"
    type="select"
  >
    <CharacterSelectItem
      character={
        Object {
          "characterId": 1,
        }
      }
      key="1"
    />
    <CharacterSelectItem
      character={
        Object {
          "characterId": 2,
        }
      }
      key="2"
    />
    <CharacterSelectItem
      character={
        Object {
          "characterId": 3,
        }
      }
      key="3"
    />
  </Input>
</FormGroup>
`;

exports[`rendering should render valid snapshot with non-default props 1`] = `
<FormGroup
  tag="div"
>
  <Label
    htmlFor="characterId"
    tag="label"
  >
    Character
  </Label>
  <Input
    data-spec="character-select-field"
    defaultValue={5}
    id="character-id"
    name="characterId"
    onChange={[Function]}
    tag="p"
    type="select"
  >
    <option
      data-spec="character-select-null-value"
      value={null}
    >
      Select a Character
    </option>
  </Input>
</FormGroup>
`;
